trieste/models/gpflow/sampler.py

```diff
@@ -0,0 +1,348 @@
+        """
+        Return approximate samples from the `model` specified at :meth:`__init__`. Multiple calls to
+        :meth:`sample`, for any given :class:`IndependentReparametrizationSampler` and ``at``, will
+        produce the exact same samples. Calls to :meth:`sample` on *different*
+        :class:`IndependentReparametrizationSampler` instances will produce different samples.
+
+        :param at: Where to sample the predictive distribution, with shape `[..., 1, D]`, for points
+            of dimension `D`.
+        :param jitter: The size of the jitter to use when stabilising the Cholesky decomposition of
+            the covariance matrix.
+        :return: The samples, of shape `[..., S, 1, L]`, where `S` is the `sample_size` and `L` is
+            the number of latent model dimensions.
+        :raise ValueError (or InvalidArgumentError): If ``at`` has an invalid shape or ``jitter``
+            is negative.
+        """
+        tf.debugging.assert_shapes([(at, [..., 1, None])])
+        tf.debugging.assert_greater_equal(jitter, 0.0)
+
+        mean, var = self._model.predict(at[..., None, :, :])  # [..., 1, 1, L], [..., 1, 1, L]
+        var = var + jitter
+
+        if not self._initialized:
+            self._eps.assign(
+                tf.random.normal([self._sample_size, tf.shape(mean)[-1]], dtype=tf.float64)
+            )  # [S, L]
+            self._initialized.assign(True)
+
+        return mean + tf.sqrt(var) * tf.cast(self._eps[:, None, :], var.dtype)  # [..., S, 1, L]
+
+
+class BatchReparametrizationSampler(ReparametrizationSampler):
+    r"""
+    This sampler employs the *reparameterization trick* to approximate batches of samples from a
+    :class:`ProbabilisticModel`\ 's predictive joint distribution as
+
+    .. math:: x \mapsto \mu(x) + \epsilon L(x)
+
+    where :math:`L` is the Cholesky factor s.t. :math:`LL^T` is the covariance, and
+    :math:`\epsilon \sim \mathcal N (0, 1)` is constant for a given sampler, thus ensuring samples
+    form a continuous curve.
+    """
+
+    def __init__(self, sample_size: int, model: ProbabilisticModel):
+        """
+        :param sample_size: The number of samples for each batch of points. Must be positive.
+        :param model: The model to sample from.
+        :raise ValueError (or InvalidArgumentError): If ``sample_size`` is not positive.
+        """
+        super().__init__(sample_size, model)
+
+        # _eps is essentially a lazy constant. It is declared and assigned an empty tensor here, and
+        # populated on the first call to sample
+        self._eps = tf.Variable(
+            tf.ones([0, 0, sample_size], dtype=tf.float64), shape=[None, None, sample_size]
+        )  # [0, 0, S]
+
+        # for some reason graph compilation is resulting in self._eps reporting the wrong shape
+        # we therefore use an extra boolean variable to keep track of whether it's initialised
+        self._initialized = tf.Variable(False)
```

Comment:
IMPORTANT: because batch_monte_carlo_expected_improvement previously only used BatchReparametrizationSampler it could reset it by setting _initialized to False. Now that it can use arbitrary samplers, we probably need to add a reset_sampler() method to the ABC and have it reset self._initialized to False when called, to ensure new samplers don't forget to do this.

Reply:
which ABC?

Reply:
Done!